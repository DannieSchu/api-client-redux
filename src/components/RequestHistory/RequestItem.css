.RequestItem {
  h3 {
    margin: 0 auto;
    text-transform: uppercase;
    color: white;
  }

  p {
    font-weight: 400;
    font-size: 0.8rem;
  }
}
