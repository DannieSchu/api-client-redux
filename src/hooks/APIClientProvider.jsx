import React, { createContext, useState, useEffect, useContext } from 'react';
import PropTypes from 'prop-types';
import { apiData } from '../services/apiData.jsx';

const APIClientContext = createContext();

export const APIClientProvider = ({ children }) => {
  const [method, setMethod] = useState('get');
  const [url, setUrl] = useState('');
  const [body, setBody] = useState('');
  const [results, setResults] = useState(null);
  const [requests, setRequests] = useState([]);
  const [loading, setLoading] = useState(false);

  useEffect(() => {
    const storage = JSON.parse(localStorage.getItem('requests'));
    if(storage) setRequests(storage);
  }, []);

  const stateFactory = {
    url: setUrl,
    method: setMethod,
    body: setBody
  };

  const handleChange = ({ target }) => stateFactory[target.name](target.value);

  const handleSubmit = e => {
    e.preventDefault();
    setLoading(true);
    apiData(url, method, body)
      .then(res => {
        setResults(res),
        setLoading(false);
      });
    setRequests(prevRequests => {
      if(prevRequests.filter(request => request.url === url && request.method === method).length < 1) {
        return [{
          url,
          method,
          body: body || null
        }, ...prevRequests];
      } else {
        return prevRequests;
      }
    });
    localStorage.setItem('requests', JSON.stringify(requests));
  };

  const handleClick = ({ url, method, body }) => {
    setUrl(url);
    setMethod(method);
    setBody(body || null);
    setResults(null);
  };

  const handleDelete = () => {
    localStorage.clear();
    setRequests([]);
  };

  return (
    <APIClientContext.Provider value={{
      url,
      method,
      body,
      results,
      requests,
      loading,
      handleChange,
      handleSubmit,
      handleClick,
      handleDelete
    }}>
      {children}
    </APIClientContext.Provider>
  );
};

APIClientProvider.propTypes = {
  children: PropTypes.node
};

export const useRequestContext = () => {
  const context = useContext(APIClientContext);
  return context;
};

export const useUrl = () => {
  const { url } = useRequestContext();
  return url;
};

export const useMethod = () => {
  const { method } = useRequestContext();
  return method;
};

export const useBody = () => {
  const { body } = useRequestContext();
  return body;
};

export const useLoading = () => {
  const { loading } = useRequestContext();
  return loading;
};

export const useResults = () => {
  const { results } = useRequestContext();
  return results;
};

export const useRequests = () => {
  const { requests } = useRequestContext();
  return requests;
};

export const useHandleChange = () => {
  const { handleChange } = useRequestContext();
  return handleChange;
};

export const useHandleSubmit = () => {
  const { handleSubmit } = useRequestContext();
  return handleSubmit;
};

export const useHandleClick = () => {
  const { handleClick } = useRequestContext();
  return handleClick;
};

export const useHandleDelete = () => {
  const { handleDelete } = useRequestContext();
  return handleDelete;
};
