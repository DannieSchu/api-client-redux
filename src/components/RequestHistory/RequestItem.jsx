import React from 'react';
import PropTypes from 'prop-types';
import styles from './RequestItem.css';
import { useUrl, useMethod, useBody, useHandleClick } from '../../hooks/APIClientProvider';

const RequestItem = () => {
  const url = useUrl();
  const method = useMethod();
  const body = useBody();
  const handleClick = useHandleClick();

  return (
    <section className={styles.RequestItem} onClick={() => handleClick({ url, method, body })}>
      <h3>{method}</h3>
      <p>{url}</p>
    </section>
  );
};

RequestItem.propTypes = {
  method: PropTypes.string,
  url: PropTypes.string,
  onClick: PropTypes.func
};

export default RequestItem;
