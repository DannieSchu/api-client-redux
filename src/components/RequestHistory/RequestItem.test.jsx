import React from 'react';
import { shallow } from 'enzyme';
import RequestItem from './RequestItem.jsx';
import { APIClientProvider } from '../../hooks/APIClientProvider.jsx';

describe('RequestItem', () => {
  it('matches a snapshot', () => {
    const wrapper = shallow(
      <APIClientProvider>
        <RequestItem />
      </APIClientProvider>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
