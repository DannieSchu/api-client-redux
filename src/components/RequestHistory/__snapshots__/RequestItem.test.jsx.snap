// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequestItem matches a snapshot 1`] = `
<ContextProvider
  value={
    Object {
      "body": "",
      "handleChange": [Function],
      "handleClick": [Function],
      "handleDelete": [Function],
      "handleSubmit": [Function],
      "loading": false,
      "method": "get",
      "requests": Array [],
      "results": null,
      "url": "",
    }
  }
>
  <RequestItem />
</ContextProvider>
`;
